## Using Python to do Technical Analysis: Find Weighted Average Minimum Price of a Stock
### The goal is to find a Weighted-Average Support Level (i.e. a floor price) across multiple time periods

In [117]:
# Import necessary libraries
import pandas as pd
import numpy as np
import yfinance as yf
from datetime import datetime, timedelta

### Fetch historical stock data from yfinance and place it in a dataframe

In [151]:
# Fetch historical stock data from yahoo based on user inputs. Default Interval is 1 day (can be 1d, 1wk, 1mo, 3mo, etc)
def fetch_stock_data(symbol, years, interval):
    stock_data = (
        yf.Ticker(symbol).history(
            start=datetime.now() - timedelta(days=365 * years), 
            end=datetime.now(), 
            interval=interval
        )
        .reset_index())  # Adds date as a column

    return pd.DataFrame(stock_data)

# Display the full data for the last 100 years
fetch_stock_data('pfe', 100, '1d')

Date       Open       High        Low      Close  \
0     1972-06-01 00:00:00-04:00   0.148763   0.151051   0.148763   0.151051   
1     1972-06-02 00:00:00-04:00   0.151051   0.151509   0.148763   0.149220   
2     1972-06-05 00:00:00-04:00   0.149220   0.149678   0.147847   0.148763   
3     1972-06-06 00:00:00-04:00   0.148762   0.152882   0.148305   0.151966   
4     1972-06-07 00:00:00-04:00   0.151966   0.151966   0.149678   0.151966   
...                         ...        ...        ...        ...        ...   
13438 2025-09-22 00:00:00-04:00  24.280001  24.799999  24.030001  24.040001   
13439 2025-09-23 00:00:00-04:00  24.100000  24.360001  24.049999  24.129999   
13440 2025-09-24 00:00:00-04:00  24.170000  24.180000  23.980000  24.090000   
13441 2025-09-25 00:00:00-04:00  24.090000  24.150000  23.580000  23.600000   
13442 2025-09-26 00:00:00-04:00  23.730000  23.809999  23.610001  23.760000   

         Volume  Dividends  Stock Splits  
0       2458771        0.0           0.0  
1       1613885        0.0           0.0  
2       2585251        0.0           0.0  
3       2347469        0.0           0.0  
4       1032077        0.0           0.0  
...         ...        ...           ...  
13438  67863300        0.0           0.0  
13439  38955200        0.0           0.0  
13440  39669400        0.0           0.0  
13441  56738300        0.0           0.0  
13442  33327100        0.0           0.0  

[13443 rows x 8 columns]

### Find a stock's minimum close price across multiple time intervals, then average them according to weights

| Time Period | Weight |
|----------|--------|
| 1 Month  | 20%    |
| 3 Months | 20%    |
| 6 Months | 15%    |
| 1 Year   | 15%    |
| 2 Years  | 10%    |
| 3 Years  | 10%    |
| 5 Years  | 5%     |
| 10 Years | 5%     |

In [147]:
# The idea is to predict the 'floor' for selling put options, with heavier weight on 
# more recent price movements while not ignoring long-term trends. 

def wgt_avg_min_close_price(symbol, periods, weights):
        
    now = datetime.now()
    
    mins = []  # To hold weighted minimums

    # Fetch 10 years of data
    all_data = fetch_stock_data(symbol, 10, '1d')
    
    # Convert to datetime and remove timezone
    all_data['Date'] = pd.to_datetime(all_data['Date']).dt.tz_localize(None)

    # For each period, calculate the cutoff date, filter data from that date onward,
    # then find the minimum close price in that slice and append the weighted value to mins
    for period, weight in zip(periods, weights):
        cutoff = now - timedelta(days=int(period * 365))
        sliced = all_data[all_data['Date'] >= cutoff]
        
        if not sliced.empty:
            min_close = sliced['Close'].min()
            mins.append(min_close * weight)
            
            # Uncomment the following line if you'd like to see the minimum values
            print(f"{round(period,2)}-Year Min: ${min_close:.2f}")  
    return sum(mins)

### Define User Inputs

In [152]:
my_symbol = "PFE"

# Time periods in years and corresponding weights
periods = [1/12, 1/4, 1/2, 1, 2, 3, 5, 10]  
weights = [0.2, 0.2, 0.15, 0.15, 0.1, 0.1, 0.05, 0.05]

wgt_avg_min = wgt_avg_min_close_price(my_symbol, periods, weights)
    
print(f"\nWgt Avg Min Close Price for {my_symbol} Stock: ${wgt_avg_min:.2f}")

0.08-Year Min: $23.60
0.25-Year Min: $23.29
0.5-Year Min: $20.83
1-Year Min: $20.83
2-Year Min: $20.83
3-Year Min: $20.83
5-Year Min: $20.83
10-Year Min: $18.10

Wgt Avg Min Close Price for PFE Stock: $21.74


### We can even get more granular and run many more time periods and weights.
### But the number of elements in each list must be the same! And the weights must sum to 100%
| Interval  | Weight |
|-----------|--------|
| Month 1   | 5%     |
| Month 2   | 5%     |
| Month 3   | 5%     |
| Month 4   | 5%     |
| Month 5   | 5%     |
| Month 6   | 5%     |
| Month 7   | 5%     |
| Month 8   | 5%     |
| Month 9   | 5%     |
| Month 10  | 5%     |
| Month 11  | 5%     |
| Month 12  | 5%     |
| Month 18  | 4%     |
| Year 2    | 4%     |
| Year 3    | 4%     |
| Year 4    | 4%     |
| Year 5    | 4%     |
| Year 6    | 4%     |
| Year 7    | 4%     |
| Year 8    | 4%     |
| Year 9    | 4%     |
| Year 10   | 4%     |


In [153]:
periods = [1/12, 2/12, 3/12, 4/12, 5/12, 6/12, 
           7/12, 8/12, 9/12, 10/12, 11/12, 12/12, 
           18/12, 2, 3, 4, 5, 6, 7, 8, 9, 10]

weights = [0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 
           0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 
           0.04, 0.04, 0.04, 0.04, 0.04, 
           0.04, 0.04, 0.04, 0.04, 0.04]
           
my_symbol = "PFE"

wgt_avg_min = wgt_avg_min_close_price(my_symbol, periods, weights)
    
print(f"\nWeighted Average Minimum Close Price for {my_symbol} Stock: ${wgt_avg_min:.2f}")

0.08-Year Min: $23.60
0.17-Year Min: $23.29
0.25-Year Min: $23.29
0.33-Year Min: $22.73
0.42-Year Min: $21.65
0.5-Year Min: $20.83
0.58-Year Min: $20.83
0.67-Year Min: $20.83
0.75-Year Min: $20.83
0.83-Year Min: $20.83
0.92-Year Min: $20.83
1.0-Year Min: $20.83
1.5-Year Min: $20.83
2-Year Min: $20.83
3-Year Min: $20.83
4-Year Min: $20.83
5-Year Min: $20.83
6-Year Min: $20.83
7-Year Min: $20.83
8-Year Min: $20.83
9-Year Min: $19.27
10-Year Min: $18.10

Weighted Average Minimum Close Price for PFE Stock: $21.18


#### As you can observe, the original weighted average minimum was 21.74
#### But after adding more time periods and weights, it decreased by more than 50 cents to 21.18
#### Furthermore, there seems to be a VERY strong support level at 20.83 (e.g. basically a strike of 21)

#### Now I can be reasonably certain the stock won't fall below 21
#### Even though it was recently that low, that's a historical low
#### The current price is around 24 per share. 
#### So 22-23 might be a pretty safe strike price to sell puts, given the floor of 21 per share

# Why This Matters
#### If I'm writing a put option, I want to be fairly certain my selected strike price is reasonably low to collect premium
 But not so low that I get assigned (e.g. I'm forced to buy the shares below market value for an unrealized loss)
#### This is one method of technical analysis that models the minimum 'floor' price based on actual data 
(We're using actual close prices and analyzing thousands of entries, e.g. we're not just guessing by looking at charts and visuals)
#### Taken alone it may be a weak indicator, but when combined with other data, it can help you make better data-driven decisions
Of course there are many other uses for this type of output, I'm simply sharing how I leverage this sort of data analysis in my own day to day trading